### DISCUSSION TITLE
Feature Request: Individual Display Configuration

### DISCUSSION THREAD

**Discussion by ghost** [input-leap/input-leap#1745] [Feature Requests]
> This issue has been migrated from old Barrier Github repository https://github.com/debauchee/barrier/issues/127
> 
> Issue created on: 2018-09-07 by @ghost
> Issue last updated on: 2021-10-09

### Configuration ###
Below is a diagram of my layout. I have an additional laptop which is the client
in the place of the "Dead Section." As barriers treats all displays as one giant
display, it can be difficult to configure a layout which is convenient for use.
The only option I have found to work is to set it directly left. This only
allows access to the Dead Section via External Display 2, but it would be nice
to access it from each respective display.

```
------------------------- -------------------------
|                       | |                       |
|      1920 x 1080      | |     1920 x 1080       |
|   External Display 2  | |   External Display 1  |
| (Cannot go down here) | |  (Cannot go diagonal) |
------------------------- -------------------------
------------------------- -------------------------
| ///////////////////// | |                       |
| /// Dead Section //// | |      1920 x 1080      |
| ///////////////////// | |     Laptop Display    |
| / (Client Position) / | | (Cannot go left here) |
------------------------- -------------------------
```

### Operating Systems ###

* Client
```
System:
  Kernel: 4.18.5-1-default x86_64 bits: 64 
  Desktop: Gnome 3.28.3 Distro: openSUSE Tumbleweed 20180831 
Machine:
  Type: Laptop System: System76 product: Lemur v: lemu6 
  Mobo: System76 model: Lemur v: lemu6 serial: <root required> 
  UEFI: American Megatrends v: 1.05.06RS76 date: 11/29/2015 
Battery:
  ID-1: BAT0 charge: 32.0 Wh condition: 32.0/45.3 Wh (71%) 
CPU:
  Topology: Dual Core model: Intel Core i7-6500U bits: 64 type: MT MCP 
  L2 cache: 4096 KiB 
  Speed: 500 MHz min/max: 400/3100 MHz Core speeds (MHz): 1: 500 2: 500 
  3: 500 4: 500 
Graphics:
  Card-1: Intel Skylake GT2 [HD Graphics 520] driver: i915 v: kernel 
  Display: x11 server: X.org 1.20.1 driver: i915 
  resolution: <xdpyinfo missing> 
  OpenGL: renderer: Mesa DRI Intel HD Graphics 520 (Skylake GT2) 
  v: 4.5 Mesa 18.1.6 
Audio:
  Card-1: Intel Sunrise Point-LP HD Audio driver: snd_hda_intel 
  Sound Server: ALSA v: k4.18.5-1-default 
Network:
  Card-1: Realtek RTL8111/8168/8411 PCI Express Gigabit Ethernet 
  driver: r8169 
  IF: enp1s0f1 state: down mac: 80:fa:5b:25:1e:d1 
  Card-2: Intel Wireless 8260 driver: iwlwifi 
  IF: wlp2s0 state: up mac: a4:34:d9:44:13:99 
Drives:
  Local Storage: total: 471.36 GiB used: 31.45 GiB (6.7%) 
  ID-1: /dev/nvme0n1 vendor: Samsung model: SSD 950 PRO 256GB 
  size: 238.47 GiB 
  ID-2: /dev/sda vendor: Crucial model: CT250MX500SSD1 size: 232.89 GiB 
Partition:
  ID-1: / size: 40.00 GiB used: 23.17 GiB (57.9%) fs: btrfs dev: /dev/dm-2 
  ID-2: /home size: 24.99 GiB used: 8.28 GiB (33.1%) fs: xfs dev: /dev/dm-3 
  ID-3: /opt size: 40.00 GiB used: 23.17 GiB (57.9%) fs: btrfs 
  dev: /dev/dm-2 
  ID-4: /tmp size: 40.00 GiB used: 23.17 GiB (57.9%) fs: btrfs 
  dev: /dev/dm-2 
  ID-5: /var size: 40.00 GiB used: 23.17 GiB (57.9%) fs: btrfs 
  dev: /dev/dm-2 
  ID-6: swap-1 size: 15.59 GiB used: 0 KiB (0.0%) fs: swap dev: /dev/dm-1 
Sensors:
  System Temperatures: cpu: 49.0 C mobo: N/A 
  Fan Speeds (RPM): N/A 
Info:
  Processes: 286 Uptime: 1h 19m Memory: 15.58 GiB used: 1.58 GiB (10.1%) 
  Shell: zsh inxi: 3.0.20 

```
* Server
```
System:
  Kernel: 4.18.5-1-default x86_64 bits: 64 
  Desktop: Gnome 3.28.3 Distro: openSUSE Tumbleweed 20180831 
Machine:
  Type: Laptop System: Dell product: Precision 5520 v: N/A 
  Mobo: Dell model: 080VVT v: X00 serial: <root required> UEFI: Dell 
  v: 1.9.4 date: 04/23/2018 
Battery:
  ID-1: BAT0 charge: 48.7 Wh condition: 48.7/56.0 Wh (87%) 
CPU:
  Topology: Quad Core model: Intel Core i7-6820HQ bits: 64 type: MT MCP 
  L2 cache: 8192 KiB 
  Speed: 800 MHz min/max: 800/3600 MHz Core speeds (MHz): 1: 800 2: 800 
  3: 800 4: 800 5: 800 6: 800 7: 800 8: 801 
Graphics:
  Card-1: Intel HD Graphics 530 driver: i915 v: kernel 
  Card-2: NVIDIA GM107GLM [Quadro M1200 Mobile] driver: nouveau v: kernel 
  Display: x11 server: X.org 1.20.1 driver: nouveau 
  resolution: <xdpyinfo missing> 
  OpenGL: renderer: Mesa DRI Intel HD Graphics 530 (Skylake GT2) 
  v: 4.5 Mesa 18.1.6 
Audio:
  Card-1: Intel CM238 HD Audio driver: snd_hda_intel 
  Card-2: Realtek type: USB driver: snd-usb-audio 
  Sound Server: ALSA v: k4.18.5-1-default 
Network:
  Card-1: Intel Wireless 8265 / 8275 driver: iwlwifi 
  IF: wlp2s0 state: up mac: 40:a3:cc:43:73:b2 
  Card-2: Realtek RTL8153 Gigabit Ethernet Adapter type: USB driver: r8152 
  IF: enp13s0u1u2 state: up speed: 1000 Mbps duplex: full 
  mac: a4:4c:c8:9c:88:2b 
  IF-ID-1: virbr0 state: down mac: 52:54:00:f7:f7:5b 
  IF-ID-2: virbr0-nic state: down mac: 52:54:00:f7:f7:5b 
Drives:
  Local Storage: total: 1.82 TiB used: 109.45 GiB (5.9%) 
  ID-1: /dev/sda vendor: Seagate model: ST2000LM007-1R8174 size: 1.82 TiB 
RAID:
  Hardware-1: Intel SATA Controller [RAID mode] driver: ahci 
Partition:
  ID-1: / size: 350.00 GiB used: 79.30 GiB (22.7%) fs: btrfs dev: /dev/dm-0 
  ID-2: /home size: 499.99 GiB used: 30.15 GiB (6.0%) fs: xfs dev: /dev/dm-2 
  ID-3: /opt size: 350.00 GiB used: 79.30 GiB (22.7%) fs: btrfs 
  dev: /dev/dm-0 
  ID-4: /tmp size: 350.00 GiB used: 79.30 GiB (22.7%) fs: btrfs 
  dev: /dev/dm-0 
  ID-5: /var size: 350.00 GiB used: 79.30 GiB (22.7%) fs: btrfs 
  dev: /dev/dm-0 
  ID-6: swap-1 size: 31.27 GiB used: 0 KiB (0.0%) fs: swap dev: /dev/dm-1 
Sensors:
  System Temperatures: cpu: 42.0 C mobo: 40.0 C 
  Fan Speeds (RPM): cpu: 2522 
Info:
  Processes: 349 Uptime: N/A Memory: 31.26 GiB used: 2.86 GiB (9.1%) 
  Shell: zsh inxi: 3.0.20 
```
### Barrier Version ###

Version: 2.1.0-RELEASE-0b2dfd80
Build Date: Monday May 14, 2018

### Steps to reproduce bug ###

Create a configuration with the layout described above, and you will see what I am talking about.

1. Create the layout as described at the top of this bug report.
2. Attempt to move through the "Dead Section."
3. Unable to connect to the client as desired.

### Other info ###

* When did the problem start to occur? See "Steps to reproduce bug."
* Is there a way to work around it? No
* Does this bug prevent you from using Barrier entirely? No

Again, this is more of a feature request to have each monitor managed independently instead of as one giant display.

---

> Commented on: 2018-09-08 by @walker0643

This would be a nice feature to have, but it's the kind of thing that won't get done until a developer that wants it has the time to write up a patch. If you would like to do so I'd be happy to take a look at the PR. Thanks!

---

> Commented on: 2018-09-13 by @chris-ault

This is what the very basis of Synergy and Barrier I thought would be about, multi screen, multi OS, multi layout configurations.  Not everyone has the exact same size screens that are only left to right, this seems like the majority of users could use this.  I as well have a client that is above my server, dual screen client and server.  Fortunately Synergy offers (0,50) configuration on each interface.

---

> Commented on: 2019-09-19 by @Karasuni

How come this feature request is closed? I'd expect this request to still be relevant even if there's no current active development.

---

> Commented on: 2020-03-23 by @kyleschieffer

+10000000000000000

---

> Commented on: 2020-05-12 by @Karasuni

> How come this feature request is closed? I'd expect this request to still be relevant even if there's no current active development.

Re-stating this comment. Open and mark as Feature Request or provide an alternative solution for feature requests.

---

> Commented on: 2020-05-12 by @shymega

Reopening, as it is worth considering. Any help big or small welcomed.

---

> Commented on: 2020-06-10 by @benedikt-bartscher

I am having trouble with my 2x2 Monitor Setup.

```
#            +--------+ +---------+
#            |win     | |desktop  |
#            |        | |         |
#            +--------+ +---------+
# +-------+  +--------+ +---------+
# |laptop |  |desktop | |desktop  |
# |       |  |        | |         |
# +-------+  +--------+ +---------+

```

```
section: links
	win:
		right = desktop(0,50)
		down  = desktop(0,50)
	desktop:
		up(0,50) = win(0,100)
		left(0,50) = win
		left(50,100) = laptop
	laptop:
		right = desktop(50,100)
end
```

I cannot move to "win" at all. Laptop and Desktop work as expected.
Individual Display Configuration would solve this problem.
I would like to help implementing this. Has anyone any starting point or tips for me? My idea is to split every Monitor to a seperate instance of barrierc.

---

> Commented on: 2020-06-15 by @benedikt-bartscher

I got it working by spawining an instance of barrierc for every Monitor with a fixed resoultion and offset. I'll try to implement this in a cleaner way now. If someone is interested in this i will publish my fork.

---

> Commented on: 2020-07-04 by @brad007

@benedikt-bartscher please do

---

> Commented on: 2020-08-11 by @shahapurkar

> <..> If someone is interested in this i will publish my fork.

Kindly Do. Thank you.

---

> Commented on: 2020-10-07 by @MathiasBaumgartinger

I do support this claim with a setup:

 [Client; Win 1080 144hz][Server; Win 1080p] [Server; Win 4k]

Log says following:

```[2020-10-07T21:07:28] INFO: starting server
[2020-10-07T21:07:28] INFO: config file: C:\...
[2020-10-07T21:07:28] INFO: log level: INFO
[2020-10-07T21:07:28] INFO: service command updated
[2020-10-07T21:07:28] INFO: got ipc shutdown message
[2020-10-07T21:07:28] NOTE: disconnecting client "Laptop"
[2020-10-07T21:07:28] NOTE: client "Laptop" has disconnected
[2020-10-07T21:07:28] NOTE: stopped server
[2020-10-07T21:07:29] INFO: process 9832 was shutdown gracefully
[2020-10-07T21:07:29] INFO: starting new process as privileged user
[2020-10-07T21:07:29] INFO: drag and drop enabled
[2020-10-07T21:07:29] ERROR: failed to get desktop path, no drop target available, error=2
started server (IPv4/IPv6), waiting for clients
server status: active
[2020-10-07T21:07:30] INFO: OpenSSL 1.0.2l  25 May 2017
[2020-10-07T21:07:30] INFO: accepted secure socket
[2020-10-07T21:07:30] INFO: AES256-GCM-SHA384       TLSv1.2 Kx=RSA      Au=RSA  Enc=AESGCM(256) Mac=AEAD
[2020-10-07T21:07:30] NOTE: accepted client connection
[2020-10-07T21:07:30] NOTE: client "Laptop" has connected

---

> Commented on: 2020-11-17 by @lijuntang

> I got it working by spawining an instance of barrierc for every Monitor with a fixed resoultion and offset. I'll try to implement this in a cleaner way now. If someone is interested in this i will publish my fork.

@benedikt-bartscher please share the fork, as this bug is a show stopper for barrier

---

> Commented on: 2021-01-23 by @theLMGN

@benedikt-bartscher please do share! my current system is currently infurating to use with Barrier, is as follows:
```
[D] [L] [D]
       [L] 
[D] = monitor connected to desktop, [L] = Monitor connected to laptop
```

---

> Commented on: 2021-01-30 by @EquusRaphanus

@benedikt-bartscher I'd love to use try out this fork; please share it! My current setup has my HTPC visually situated between my two computer monitors, and I'd love to make this seamless!
```
Current Barrier Setup:
[  TV  ]
[PC][PC]

Actual Setup/Ideal Barrier Setup:
[PC][TV][PC]
```


---

> Commented on: 2021-03-05 by @derWombie

@benedikt-bartscher
Would also be highly interested in this fork/solution! :-) THX

---

> Commented on: 2021-03-11 by @Delwin9999

I would also be interested in the fork.  I'm running like this:

```
|    |              |   |
| L  ||  L  ||  W  || W |
|    |    |  W  |   |   |
```

The left two are on the linux box, the right two and the lower central are all connected to the windows box.  The outer two are portrait orientation and the inner three are landscape orientation.

---

> Commented on: 2021-04-19 by @aejay-rs

Just chiming in with my own arrangement:
```
-------- ------------ --------
|      | |    PC    | |      |
|      | |          | |      |
|  PC  | ------------ |  PC  |
|      | |  Laptop  | |      |
|      | |          | |      |
-------- ------------ --------
```
Where both PC and laptop are (currently) windows.

Configured this way, out of the box, things work, but it's a bit unintuitive. I can move upward from the laptop screen to the central PC screen, but not to the side monitors. Likewise, I can move downward from the side monitors (to the top of the laptop screen, oddly enough), but I cannot move from the central PC screen to the laptop screen.

---

> Commented on: 2021-04-26 by @mura94

Would **love** for this to be supported. I'm currently working with a 1440p in landscape and a 1080p in portrait for a code screen. It works if I use my laptop as the server, but not as the client.

---

> Commented on: 2021-05-01 by @Silmakhor

I'm getting a new Imac with M1 chip. Stupidly, these don't support more than one external monitor. So I would like to use Barrier as a way to use my old mac as an external display. 

Here's what I will have:

Imac **2021** connected to XPpen graphics display **tablet**
Imac **2011** connected to **Dell** 24" external monitor.

_Ideally arranged as follows:_
Dell | 2021 | 2011
. . . . tablet . . . . . 
  
For this to work intuitively, I would need to be able to configure the Dell as a separate display in Barrier.





---

> Commented on: 2021-09-16 by @zhenruiliao

I am also affected by this issue, and unfortunately I suspect this software is unusable to myself and a significant number of other users as long as barrier does not support multiple monitors :cry: 
But I also understand the difficult of automatically calculating the appropriate switch region for arbitrary monitor configurations. I wonder if there is a compromise solution--basically allow the user to manually specify the coordinates where they would like the transition to happen in the config, and provide some tool for displaying the coordinates of the user's cursor at any given time under their current configuration (such as xdotool getmouselocation in Linux) so they can measure and calibrate on their own. 

---

> Commented on: 2021-09-22 by @voidpumpkin

Is there a way to fund the issue? Would love to push this more.

---

> Commented on: 2021-10-09 by @c-palmer

This would be such a useful feature

**Comment by Benzolio** [comment]
@benedikt-bartscher Did you end up publishing your fork with the per-monitor instances of barrierc? I would really like to try it.


My physical layout is like:
```
+----------------+ +-------+ +--------+ +--------+
|                | |Laptop | |PC mon4 | | PC mon5|
|    Manjaro     | +-------+ +--------+ +--------+
|      box       | +-------+ +--------+ +--------+
|                | |PC mon1| |PC mon2 | | PC mon3|
+----------------+ +-------+ +--------+ +--------+
```
and from my .sgc config file:
```
section: links
	ManjaroBox:
		right(0, 50) = Laptop
		right(50, 100) = PC(50, 100)
	Laptop:
		down(0, 100) = PC(0, 25)
		right = PC(0, 50)
		left = ManjaroBox(0, 50)
	PC:
		left(0, 50) = Laptop(0, 50)
		left(50, 90) = ManjaroBox(50, 90) /* 90 instead of 100 because I wanted a lower left dead corner on the PC */
		up(0, 33.33) = Laptop(0,100)
end
```
I can move left from Laptop to ManjaroBox, and I can move back down to the PC from anywhere on the bottom of the laptop screen to PC monitor 1, BUT I cannot go back up from PC monitor 1, and I cannot go right from the Laptop to PC monitor 4 (no matter where on the right edge of Laptop, the cursor goes to 50,0 of PC desktop left edge, which is the top left corner of monitor 1 instead), nor left from PC monitor 4 to Laptop (mouse just stops on left edge of monitor 4). Strangely, I can go up from the top left corner of PC monitor 4 to bottom right corner of Laptop, and I suspect this is due to some imprecision around mouse velocity and that 33.33, but that seems like the other issue mentioned in [#944](https://github.com/input-leap/input-leap/issues/944#issue-1045153071)

So it currently behaves kind of like:
```
+----------------+    +------+          
|                |<-->|Laptop|          
|    Manjaro     |    +------+<---\         
|      box       |     down \-\    \---\+-------++-------+
|                |<-\     only \-\      |PC mon4||PC mon5|
+----------------+   \-\          v     +-------++-------+
                        \-\    +-------++-------++-------+
                           \-->|PC mon1||PC mon2||PC mon3|
                               +-------++-------++-------+
```

It appears that crossing a gap in a desktop upon entering from the top is handled fine, but triggering from a desktop at the edge of a gap doesn't happen. For a horizontal move, it appears that the y coordinate is being shifted to avoid the gap, instead of offsetting the X coordinate to cross it. So I'm guessing there is logic to handle crossing vertical gaps and it is being applied to horizontal moves without switching coordinates. If so, I suspect that there is an easy fix for that. (I suspect detecting the gap edge from within the desktop is likely to be more involved.)

 I'm using it as is since it does allow me to get to the laptop without wrapping the mouse clockwise through the ManjaroBox screen, but it would be nice to have screen switching across any gaps in a desktop rectangle where there is no monitor working horizontally as well as vertically and bidirectionally in both cases.
 
Even if it isn't automated to detect the gap, I wouldn't mind defining the gap somehow in the links section of the configuration file manually, if that were an option. But, that's probably a whole new feature request.
